pontoon/translate/static/js/translate.es6

```diff
@@ -0,0 +1,499 @@
+        return translation;
+      }
+    }
+
+    return null;
+  }
+
+  /**
+   * Return the current translation status.
+   */
+  get status() {
+    if (this.approvedTranslation) {
+      return 'approved';
+    } else if (this.translations.some((t) => t.fuzzy)) {
+      return 'fuzzy'
+    } else if (this.translations.length > 0) {
+      return 'translated';
+    } else {
+      return '';
+    }
+  }
+
+  updateTranslation(string, localeCode) {
+    let url = `/entity/${this.pk}/translations/${localeCode}/`;
+    return $.post(url, {string: string}).then((translations) => {
+      return $.extend(this, {translations: translations})
+    });
+  }
+
+  /**
+   * Fetch all the entities for the given project and locale.
+   */
+  static fetchAll(projectSlug, localeCode) {
+    let url = `/project/${projectSlug}/locale/${localeCode}/entities`;
+    return $.get(url).then((entities) => {
+      return entities.map((data) => new Entity(data));
+    });
+  }
+}
+
+
+/**
+ * Base class for React Components.
+ */
+class PontoonComponent extends React.Component {
+  /** Shorthand for getting a jQuery object for a ref. */
+  $ref(refName) {
+    return $(React.findDOMNode(this.refs[refName]));
+  }
+}
+
+
+/**
+ * Root Component for the entire translation editor.
+ */
+class TranslationEditor extends PontoonComponent {
+  constructor(props) {
+    super(props)
+    this.state = {
+      loaded: false,
+      sidebarOpen: true,  // Temporarily defaults to open for dev.
+      entities: [],
+      selectedIndex: 0,
+    };
+  }
+
+  componentDidMount() {
+    Entity.fetchAll(this.props.project.slug, this.props.locale.code).then((entities) => {
+      this.setState({
+        loaded: true,
+        entities: entities,
+      });
+    });
+  }
+
+  render() {
+    if (!this.state.loaded) {
+      return <LoadingScreen />;
+    }
+
+    return (
+      <div id="translation-editor">
+        <aside id="sidebar"
+               className={classNames('advanced', {open: this.state.sidebarOpen})}
+               ref="sidebar">
+          <EntityList entities={this.state.entities}
+                      project={this.props.project}
+                      selectedIndex={this.state.selectedIndex}
+                      onSelectEntity={this.selectEntity.bind(this)} />
+          <EntityDetails entity={this.selectedEntity}
+                         onSelectNextEntity={this.selectNextEntity.bind(this)}
+                         onSelectPreviousEntity={this.selectPreviousEntity.bind(this)}
+                         onSave={this.saveTranslation.bind(this)} />
+          <div id="drag" draggable="true"></div>
+        </aside>
+      </div>
+    );
+  }
+
+  get selectedEntity() {
+    return this.state.entities[this.state.selectedIndex];
+  }
+
+  selectEntity(entityKey) {
+    let index = this.state.entities.findIndex((e) => e.pk === entityKey);
+    if (index !== -1) {
+      this.setState({
+        selectedIndex: index,
+      });
+    }
```

Comment:
Maybe. It's a bug in the client-side code if that happens, perhaps show throw and then catch and show a generic "something went wrong" message?